File: 141-linked-list-cycle/141. Linked List Cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        
        if not head:
            return False
            
        curr1 = head
        curr2 = head.next

        while curr1 and curr2 and curr1 != curr2:
            curr1 = curr1.next
            curr2 = curr2.next
            if curr2:
                curr2 = curr2.next
        
        return curr1 == curr2
        

File: 141-linked-list-cycle/141. Linked List Cycle.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
bool hasCycle(struct ListNode *head) {

    if(!head){
        return false;
    }

    if(!head->next){
        return false;
    }

    struct ListNode* slow = head;
    struct ListNode* fast = head->next;

    while(fast->next && fast->next->next){
        if (slow == fast){
            return true;
        }

        fast = fast->next->next;
        slow = slow->next;
    }

    return false;

}